catfs/db/database_badger.go

```diff
@@ -56,17 +55,95 @@ func NewBadgerDatabase(path string) (*BadgerDatabase, error) {
+	opts := bdb.db.Opts()
+	// The logic is quite convoluted:
+	// It is not enough to ask for GC, badger DB needs to run some statistic
+	// after last modification in DB (about a minute), if we don't wait
+	// the GC call will be waisted in we might claim that there is nothing to GC.
+	// Also to do GC correctly we need to Close DB before calling GC
+	// (yep, badger is crazy in this regard).
+	// Then we reopen DB and mark cleanable with GC,
+	// then on Close() true GC happens (HDD space reclaimed),
+	// and we need to reopen again.
+	// Consequently we have a timer (to give time for stats update) and we needGC flag
+	// which indicates that some updates were done on DB so we do not trash
+	// underlying FS.
+	// TODO: to do it really correct we need a queue which holds last DB modification
+	// times. But it all become very convoluted since other processes mess with DB
+	// Or repinner always does something in background and updates DB.
+	if time.Now().Before(bdb.gcNeededAt) || !bdb.needsGC {
+		// we are waiting for timer but also for updates
+		// if there are no updated in DB, timer does not kick in
+		log.Debugf("badger DB in %s stats are not updated yet, no need for GC", opts.Dir)
+		return nil
+	}
+	bdb.mu.Lock()
+	defer bdb.mu.Unlock()
+	defer func() {
+		bdb.gcNeededAt = time.Now()
+		bdb.needsGC = false
+	}()
+	var errGC error
+	var beforeGC uint64 = 0
+	var total uint64 = 0
+	var err error
+	log.Infof("Performin GC for badger DB in %s", opts.Dir)
+	for errGC == nil {
+		// At the DB opening, a new vlog is created
+		// and its size added to vlog size even if it is empty
+		// The closing truncates vlog files so we get a better read of sizes
+		err = bdb.db.Close() // GC is claimed only on close :(
+		if err != nil {
+			log.Fatalf("Could not close DB in %s", opts.Dir)
+			return err
+		}
+		bdb.db, err = badger.Open(opts)
+		if err != nil {
+			log.Fatalf("Could not reopen DB in %s", opts.Dir)
+			return err
+		}
+		// Badger people use os.FileInfo to get total size of lsm and vlog files
+		// This is not actual but allocated of hdd size
+		lsm, vlog := bdb.db.Size() // values could be stale even after Close/Open cycle, badger updates it once per minute
+		// TODO: it might be simple just to count HDD allocation ourself as it done in badger internals
+		//       for now I leave it alone, since it mostly for info messages
+		newTotal := uint64(lsm) + uint64(vlog)
+		log.Debugf("DB sizes in %s: lsm %s and vlog %s, total %s", opts.Dir, humanize.Bytes(uint64(lsm)), humanize.Bytes(uint64(vlog)), humanize.Bytes(uint64(newTotal)))
+		if beforeGC == 0 {
+			beforeGC = newTotal
+		}
+		if total != 0 {
+
+			log.Infof("At this iteration DB size in %s decreased by %s", opts.Dir, humanize.Bytes(total-newTotal))
+		}
+		total = newTotal
+
+		if errGC = bdb.db.RunValueLogGC(0.5); errGC != nil {
+			if errGC == badger.ErrNoRewrite {
+				log.Debugf("badger DB in %s has nothing for GC",opts.Dir)
+			}
+			continue
+		}
```

Comment:
What's the purpose of still running ``RunValueLogGC(0.5)`` after Close? Is it essential to call?

Reply:
According to the Note in https://github.com/dgraph-io/badger/issues/767#issuecomment-485713746
the compactors,  which collect statistics for GC, are run only on DB `Close()` at least on small size DB.
So `Close` is needed for our DB size.

But your comment made me do more digging. So some things about `badger` DB GC collector was wrongly reversed engineered by me: `RunValueLogGC()` actually compacts on the fly, no need to `Close()` to shrink the DB. The problem that `RunValueLogGC` does not collect stats by itself, but uses prepared by compactors.
Which leads for necessity of one `Close()` to make sure that compactors are forced. Apparently, even `Close`
does not force it and need about 1 minute delay between DB modification and `Close` for compactors to kick in.

So new 3cc667c6 patch get rid of cycled Close/Open (it does it only once per `runGC` call), also I get rid
of the timer since the oldest DB flush and the modification flag.

If only there were reliable way to check if DB was modified, then we would not do unnecessary `Close()`. Though I noticed that `filesystem GC` constantly messing with the metadata DB even on idle `brig`.

Reply:
Thanks for the explanation. Code looks better to me now.

One simple to gather statistic we could display would be to display how long the GC run took. I'm a little worried that it blocks all of brigs operation for considerable time, even if there weren't many writes in that time period.